purple/core.py

```diff
@@ -26,12 +26,39 @@ def __init__(self, keyboard, layout, *extras):
         self.layer_index = 0
         self._one_shot_key_buffer = []
         self._lock_key_buffer = []
+        # buffer for keys pressed
+        self._pressed_keys = []
+        # buffer for keys held
+        self._held_keys = []
         self._held_mouse_movement = ()
         self._current_time = supervisor.ticks_ms()
         self._prior_time = self._current_time
         self._prior_mouse_time = self._current_time
         self._prior_change_time = self._current_time
 
+    def _press_key(self, *keycodes):
+        """
+        Presses a key on the keyboard, and adds it to the list of pressed keys.
+        This way we can track which keys need to be released later.
+        """
+        self._hid_keyboard.press(*keycodes)
+        self._pressed_keys.append(*keycodes)
+
+    def _release_key(self, keycode):
+        """
+        Releases a key on the keyboard, and removes it from the list of pressed keys.
+        """
+        self._hid_keyboard.release(keycode)
+        self._pressed_keys.remove(keycode)
+
+    def _release_keys(self):
+        """
+        Releases all keys that are currently pressed.
+        """
+        for keycode in self._pressed_keys:
+            self._hid_keyboard.release(keycode)
```

Comment:
when keys are released they should also be removed from `self._held_keys`, otherwise if you have `Alt` held and you tap `Alt`, the computer sees a key release for `Alt` but the keyboard still thinks it's being held